### ISSUE TITLE
Bookmarks data too large for localStorage

### ISSUE THREAD

**Issue by joeytwiddle** [xbrowsersync/app#20]
Apologies, this is three different bugreports in three days!

Running the store extension on my Mac, I don't have any problem with corrupt bookmarks. But I do exceed storage capacity!

```
xBrowserSync.App.Background:syncBookmarks, WARNING: Error syncing bookmarks; {"stack": "...

Error: Failed to execute 'setItem' on 'Storage': Setting the value of 'xBrowserSync-cachedBookmarks' exceeded the quota.
    at Object.setInLocalStorage [as Set] (chrome-extension://lcbjdhceifofjlpecfpeimnnphbcjgnc/js/background.min.js:2732:22)
    at Object.Set (chrome-extension://lcbjdhceifofjlpecfpeimnnphbcjgnc/js/background.min.js:389:43)
    at getCachedBookmarks (chrome-extension://lcbjdhceifofjlpecfpeimnnphbcjgnc/js/background.min.js:1613:37)
    at refreshCachedBookmarks (chrome-extension://lcbjdhceifofjlpecfpeimnnphbcjgnc/js/background.min.js:1537:16)
    at chrome-extension://lcbjdhceifofjlpecfpeimnnphbcjgnc/js/background.min.js:2068:13
    at ...

..."}
```

(I have around 14,900 bookmarks, 6MB when exported as HTML, and my own API server with the limit set to 100MB for testing!)

I have heard people recommending to avoid `localStorage` when possible, because it is synchronous and is has limits (around 2MB?) in some browsers.

People recommend using something like IndexedDB instead (or ServiceWorkers Cache API). Although that's for front-end code, I'm not so sure about extensions.

**Comment by nero120** [comment]
Don't be sorry @joeytwiddle, your efforts are definitely appreciated! Yes this is definitely a limitation that needs to be addressed, `indexedDB` seems to make the most sense as a replacement cache store but I'll need to consider how that will affect the mobile apps as it's unified code.

**Comment by nero120** [comment]
Just a heads up, this should be slightly mitigated in v1.4.0 which now compresses the data before it is encrypted, typically reducing the size of the encrypted data by around 40-60%. As the bookmarks data is cached in local storage in encrypted form, you should be able to store far more bookmarks before hitting the local storage quota. 😄

**Comment by nero120** [comment]
This will be resolved by adding the `unlimitedStorage` manifest permission once `storage.local` API has been implemented (see #38).

**Comment by nero120** [comment]
This is resolved by d28125b7e1a29b72f6186ea0c9c4dab2c0c8b8e6 and 7147c0244c867bf0a2aa67a2120a5c85b1923ef1 adding the `unlimitedStorage` manifest permission.
